ENH: Add query and query_bulk to sindex

adriangb opened the pull request:
The next step in integration of `pygeos.strtree`.

A couple of notes:
1. I included sorting for now, but am happy to remove it if it adds too much complexity.
2. I included support for other predicates (and tests, it works) but will be happy to remove it for now if we want to do that separately.

A general comment:
I think it would be cool to have a spatial index abstract base class that lives inside of `sindex.py` that starts to sketch out what API we expect for spatial indexes. Any thoughts on that are welcome.

CCing @jorisvandenbossche  @martinfleis  @brendan-ward

adriangb commented on geopandas/tests/test_sindex.py:
```diff
@@ -1,13 +1,15 @@
 from numpy.testing import assert_array_equal
 
 import geopandas
+from geopandas.array import _shapely_to_geom as shapely_to_geom
```
@jorisvandenbossche I'd like to get your opinion on this. Generally, I like to respect the use of `_`, but I also don't think it's ideal to re-implement this here just to avoid the import.

jorisvandenbossche replied:
I think that should be handled inside the tree `query` method (ensuring that an input of shapely geometry gets converted to a pygeos geometry). And then in the tests here you can just use shapely's box.

jorisvandenbossche replied:
> I assume we're okay not checking in bulk_query?

Yes, that should be OK.

adriangb replied:
Cool, change implemented then.

martinfleis commented on geopandas/sindex.py:
```diff
@@ -81,22 +223,119 @@ def is_empty(self):
 if compat.HAS_PYGEOS:
 
-    from pygeos import STRtree, box, points  # noqa
+    import geopandas
```
Just a quick note, (I'll do a proper look later), can you import only classes you need from their files instead of whole geopandas?

adriangb replied:
I was running into circular imports with `geopandas.geoseries.GeoSeries`. We can either import it as `import geopandas.geoseries` here or I think we could also change `from .sindex import get_sindex_class, has_sindex` to `import sindex` in `geopandas.base` instead which would allow `from geopandas import GeoSeries`.

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -70,6 +89,129 @@ def __init__(self, geometry):
+                ]
+
+            if not sort:
+                return np.array(tree_query)  # unsorted
```
```suggestion
                return np.array(tree_query, dtype=np.intp)  # unsorted
```

? You do that for the empty arrays.

(and the same for the other occurrences that create an array)

adriangb replied:
Good catch, will do.

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -81,22 +223,120 @@ def is_empty(self):
+                        "Got `geometry` of type `%s`, " % type(geometry)
+                        + "`geometry` must be a GeoSeries, GeometryArray, "
+                        "numpy.ndarray or other iterable."
+                    )
```
instead of this try/except, I think it is fine to simply convert anything else (which is not a GeoSeries or GeometryArray) to a ndarray with `np.asarray(geometry)`

adriangb replied:
That's a good idea. That should take care of most iterables.

adriangb replied:
Implemented

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -115,11 +355,12 @@ def intersection(self, coordinates, objects=False):
                     "Invalid coordinates, must be iterable in format "
-                    "(minx, miny, maxx, maxy) (for bounds) or (x, y) (for points)."
+                    "(minx, miny, maxx, maxy) (for bounds) or (x, y) (for points). "
+                    "Got `coordinates` = %s." % coordinates
```
Can you use `format()` instead of `%` ?

adriangb replied:
Of course!

I'm a big f-string user, but when I forget what Python X.Y supports, so I just default to `%` when in doubt. We are >=3.5 here so yeah `format` is much better.

adriangb replied:
Fixed

adriangb commented on geopandas/sindex.py:
```diff
@@ -2,10 +2,22 @@
 from . import _compat as compat
 
 
+VALID_QUERY_PREDICATES = {
+    None,
+    "intersects",
+    "within",
+    "contains",
+    "overlaps",
+    "crosses",
+    "touches",
+}
+
+
```
I think this could be a good first step towards defining what we expect a generic `sindex` API to look like. `sindex` classes can adopt these defaults or a custom set of supported predicates, and what I expect will eventually be generic/abstract base class code will catch errors appropriately.

jorisvandenbossche replied:
Let's keep the discussion about a generic base class in https://github.com/geopandas/geopandas/issues/1344

adriangb commented on geopandas/sindex.py:
```diff
@@ -81,22 +223,119 @@ def is_empty(self):
 
+        # set of valid predicates for this spatial index
+        # by default, the global set
+        valid_query_predicates = VALID_QUERY_PREDICATES
```
See [above comment](https://github.com/geopandas/geopandas/pull/1401/files/1b3feaa15560de99a103d63d2df13359c1216bf0#r417096495).

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -225,11 +239,16 @@ def is_empty(self):
     from . import geoseries  # noqa
     from .array import GeometryArray  # noqa
-    from pygeos import STRtree, box, points, Geometry  # noqa
+    from pygeos import STRtree, box, points, Geometry, from_shapely, from_wkb  # noqa
```
Can you do `import pygeos`, and then `pygeos.from_shapel(..)` etc? Since we also have `from_shapely` and `from_wkb` functions in geopandas, this gets confusing in the code to know which one it is

adriangb replied:
Sure!

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -70,6 +89,129 @@ def __init__(self, geometry):
+                if predicate == "within":
+                    # since these are inverse, we can flip the operation
+                    # and test with prepared predicates from tree
+                    predicate = "contains"
```
Shouldn't the geometries get flipped in order as well? (because it seems you are now using the same order for both contains and within ?)

adriangb replied:
The logic here is:
```python3
if "within":
    tree[i].contains(geometry)
if "contains":
   geometry.contains(tree[i])
```

So indeed, the geometries do get flipped. What gets appended to the results is `i`, which in both cases is the index of the geometry in the tree. `query` only outputs indexes for tree geometries.

adriangb replied:
I will try to add more comments to document this, if there is any specific way I can clarify this please let me know.

adriangb replied:
I re-organized this section and added some more comments. Please let me know if it is clear now.

jorisvandenbossche replied:
Yeah, I missed they were using a different order in the actual call (I actually assumed this block was for within and contains, but it was actually for within and intersects, and contains was done below differently).

But added some new comments now ;)

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -70,6 +89,129 @@ def __init__(self, geometry):
+            ----------
+            geometry : shapely geometry
+                A single shapely geometry to query against the spatial index.
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
```
```suggestion
            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', \
'crosses', 'touches'}, optional
```

That's a (somewhat ugly) way to solve the "too long line length" (instead of adding `# noqa: E501`)

adriangb replied:
I think that came from `black` or something, I _have_ `#noqa: E501` there. But I'll fix it.

adriangb replied:
Okay, no problem. I switched to `\` and fixed all of the other long lines.

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -81,22 +237,137 @@ def is_empty(self):
+                )
+
+            if isinstance(geometry, BaseGeometry):
+                # handle shapely geometries
+                if compat.PYGEOS_SHAPELY_COMPAT:
+                    geometry = from_shapely(geometry)
+                # fallback going through WKB
+                elif geometry.is_empty and geometry.geom_type == "Point":
+                    # empty point does not roundtrip through WKB
+                    geometry = from_wkb("POINT EMPTY")
+                else:
+                    geometry = from_wkb(geometry.wkb)
```
Since this is duplicating the code in `_shapely_to_geom` / `_shapely_to_pygeos`, it is fine to use those directly instead of duplicating the code. 
(yes, those were coded as "private", but it's fine to use them for now within geopandas)

adriangb replied:
Ok, with that okay I went ahead and imported `_shapely_to_geom` (we don't really need `_shapely_to_pygeos`).

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+                        predicate, self.valid_query_predicates
+                    )
+                )
+            # handle empty / invalid geometries
```
minor nit: a bit more space ("breathing room") would be good here.  Consider a newline before each conditional block or comment preceding a conditional.

adriangb replied:
Added some breathing room 👍

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+                )
+            # handle empty / invalid geometries
+            if geometry is None:
+                # this is the behavior in pygeos.strtree.query, we mimic it here
```
suggestion:
```
# return an empty integer array, similar to pygeys.STRtree.query.
```

adriangb replied:
Implemented!

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+            if geometry.is_empty:
+                return np.array([], dtype=np.intp)
+
+            # get bounds
```
comment is unnecessary

adriangb replied:
Removed comment.

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+
+            # get bounds
+            bounds = geometry.bounds  # rtree operates on bounds
+            tree_query = list(self.intersection(bounds, objects=False))
```
Readability suggestion: "query" implies an operation, but "tree_query" here is a list of results.  "results" or "tree_idx" would probably be better here.

adriangb replied:
Agreed and implemented.

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+                    if getattr(geometry, predicate)(self._geometries[i])
+                ]
+
+            if not sort:
```
recommend flipping this conditional with the default return.  The default value for `sort` is `False`, so the default (last) return should be too.

adriangb replied:
Agreed and implemented.

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+                    if self._prepared_geometries[i] is None:
+                        # if not already prepared, prepare and cache
+                        self._prepared_geometries[i] = prep(self._geometries[i])
+                    if getattr(self._prepared_geometries[i], predicate)(geometry):
```
Use of `getattr` masks the operation a bit, and use of "intersects" here (left side tree geom rather than input geom) is not immediately obvious without further explanation.  

Recommend adding a comment like
```
# Test the prepared tree geometry against the input geometry.
# For "intersects" this leverages the cache of prepared tree geometries.
```

adriangb replied:
I've added a lot more comments to this section from the feedback. Please let me know if it seems more or less clear.

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+                    if getattr(self._prepared_geometries[i], predicate)(geometry):
+                        res.append(i)
+                tree_query = res
+            elif predicate == "contains" and len(tree_query) > 1:
```
Limiting this case to `len(tree_query) > 1` is not immediately obvious.  Am I correct in thinking that even if there is 1 hit from the tree, it does not guarantee that the input geometry `contains` the tree geometry?  As in, their bounding boxes intersect, but without evaluating the predicate, we don't know how the actual geometries relate to each other.

adriangb replied:
This is a good catch. The way it is written right now, it's a bug. I guess my tests are not very good (I'll add one for this specific case). I'm going to rewrite as follows:

```python3
elif predicate == "contains":
    if len(tree_idx) > MIN_MATCHES_FOR_PREP:
        # prepare this geometry
        geometry = prep(geometry)
     tree_idx = [
         i
         for i in tree_idx
         if getattr(geometry, predicate)(self._geometries[i])
     ]
```

As an aside, what I just called `MIN_MATCHES_FOR_PREP` above is currently just `1`. Do you think this is a check we should do? Is there a better number? Or should we always prep?

adriangb replied:
@jorisvandenbossche answered below. We are going to always use prepared geometries.

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
+                If predicate is provided, a prepared version of the input geometry is tested using
+                the predicate function against each item in the index whose extent intersects the
+                envelope of the input geometry: predicate(geometry, tree_geometry).
```
This isn't technically correct, per below implementation.

```
If predicate is provided, the input geometry is tested using the predicate function against each item in the index whose extent intersects the envelope of the input geometry: predicate(geometry, tree_geometry).  If possible,
prepared geometries are used to help speed up the predicate operation.
```
(sorry for bad prior suggestion here)

adriangb replied:
Thanks. I am going to leave any mention of prepared geometries out of the pygeos sindex docstrings since it currently does not use prepared geometries and if it does, that might happen internally within pygeos.

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+            Returns
+            -------
+            ndarray with shape (2, n)
+                The first subarray contains input geometry indexes.
```
suggest adding `integer indexes` to be really clear here, since inputs include indexed types (`GeoSeries`).

adriangb replied:
Agreed

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+            """  # noqa: E501
+            # Iterates over geometry, applying func.
+            tree_index = []
+            geo_index = []
```
minor nit: suggest `geom_index` and use of `geom` over `geo` here and below.  ("geo" has overloaded meaning that isn't always a geometry)

adriangb replied:
I changed it to `input_geometry_index` to be even more explicit.

brendan-ward commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+
+            for i, geo in enumerate(geometry):
+                res = self.query(geo, predicate=predicate, sort=sort)
+                if res.size > 0:
```
This conditional seems unnecessary; if `res` is length 0, then the following lines have no effect.
The following comment is unclear; there doesn't appear to be any sorting in the following lines.

adriangb replied:
You are right. I removed the conditional and comment.

adriangb commented:
> @adriangb sorry for the slow rate of my review here; I've only made it through the Rtree implementation so far.

No worries, I know that I've been putting a lot of stuff up for review!

> I don't know if it would lead to more proliferation of files than is desirable here, but it struck me that the Rtree and STRtree implementations could be in separate files, to keep their imports cleaner? Maybe making `sindex` into a package instead of a single file...

I think that loops back to the discussion in #1344. I see keeping them in the same file as the "default" option. If a consensus is reached on an alternative, I'd be happy to implement it.

martinfleis commented:
>>I don't know if it would lead to more proliferation of files than is desirable here, but it struck me that the Rtree and STRtree implementations could be in separate files, to keep their imports cleaner? Maybe making sindex into a package instead of a single file...

> I think that loops back to the discussion in #1344. I see keeping them in the same file as the "default" option. If a consensus is reached on an alternative, I'd be happy to implement it.

I would keep it as it is for now, expecting that major changes will happen in future.

jorisvandenbossche commented:
Yes, let's keep it in the current file for now, to minimize the diff.

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,150 @@ def __init__(self, geometry):
+
+            # Check predicate
+            # When possible, we want to use cached prepared geometries
+            # from the tree to compare tree_geom.predicate(input_geom)
```
The other way around, right? (`input_geom.predicate(tree_geom)`) 
At least, that's what I think pygeos does? (from its docstring: `predicate(geometry, tree_geometry)`), and also what you put in a comment below

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,150 @@ def __init__(self, geometry):
+            # Since only certain predicates support prepared geometries,
+            # the fallback is to check non-prepared geometries.
+            if predicate in ("intersects", "within"):
+                # For these two predicates, we compare tree_geom.predicate(input_geom)
```
For "within" I understand now why we are flipping this around. But why do this for "intersects"? I think it will be more efficient to prepare the input geometry, since this is called multiple times

adriangb replied:
Well we can cache the tree geometry, but not the input geometries. So I guess it depends on the specific dataset. I did it this way so that we are caching whenever possible, but I'm open to doing it the other way around.

jorisvandenbossche replied:
Assuming a single call to query_bulk, each geometry will be passed once to query, so I think it is OK they are not cached. 

(we currently also don't do this, and caching prepared geoms in general is a separate topic we can discuss with pygeos, I would say. We shouldn't put too much effort in optimizing the current shapely code, at least not beyond what we have now. As it will become obsolete soon)

jorisvandenbossche replied:
(BTW, this is also how it is done in pygeos now (prepare the input geometry, not the tree geometry)

adriangb replied:
Okay, no problem, I can move intersects to the other loop so that the input geometry is prepared. Do we still want to cache the prepared tree geometry to "within", or should I get rid of that too?

jorisvandenbossche replied:
Keeping it for "within" might be fine. We do switch the order in `sjoin`, so in practice are using prepared geoms there.

adriangb replied:
Right, I got the idea from sjoin. Once we put `bulk_query` into sjoin, my idea would be to remove prepared geometries from sjoin (they're not compatible with pygeos).

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -70,6 +91,150 @@ def __init__(self, geometry):
+            elif predicate is not None:
+                # For the remaining predicates,
+                # we compare input_geom.predicate(tree_geom)
+                if len(tree_idx) > 1 and predicate == "contains":
```
```suggestion
                if predicate == "contains":
```

I don't think this check is needed. Even for a single predicate call, preparing the geom can already be beneficial

adriangb replied:
Okay, thank you for confirming that. I wasn't sure if the threshold for which preping was faster was 1 comparison, 5 comparisons or 10 comparisons. If prepping is faster for even 1 comparison I agree, the always prep is better.

jorisvandenbossche commented on geopandas/tests/test_sindex.py:
```diff
@@ -172,10 +174,309 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+    def test_query_empty_geometry(self, test_geom, expected_error, expected_value):
+        """Tests the `query` method with empty geometry.
+        """
+        if expected_error is not None:
```
Is there going to be a case where you actually expect an error (now this is always None), otherwise this can be cleaned up a bit?

adriangb replied:
I think that can be removed in some tests, I copy pasted and wasn't sure if it would be needed or not. I'll go through and remove where it is not being used, I guess it can always be added back later.

adriangb replied:
Parameter removed in next commit.

jorisvandenbossche commented on geopandas/tests/test_sindex.py:
```diff
@@ -172,10 +174,309 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+
+    @pytest.mark.parametrize(
+        "predicate, test_geom, expected", (("test", (-1, -1, -0.5, -0.5), [[], []]),),
+    )
```
If there is only a single value for the parameters, using parametrize is maybe not needed? (or do you anticipate to add more here?)

adriangb replied:
I kind of like having it there, that way it is easy to tell what the input parameters are, vs. hardcoding it in the test.

jorisvandenbossche replied:
OK, fine for me

jorisvandenbossche commented on geopandas/tests/test_sindex.py:
```diff
@@ -172,10 +174,309 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+        # determine if it should use shapely or pygeos geometry objects
+        # note: for this test, test_geoms (note plural) is a list already
+        test_geoms = geopandas.GeoSeries(test_geoms, index=range(len(test_geoms)))
+        if expected_error is not None:
```
same comment here about the expected error

adriangb replied:
Done, parameter removed.

adriangb commented:
Thank you the review @jorisvandenbossche!

Aside from the formatting things and unused parameters in tests, the main issue (which had also been brought up by @brendan-ward) was the logic in prepping geometries and such for the `rtree` `query`.

I think I've simplified it now and as discussed we are now always comparing `input_geom.predicate(tree_geom)` except for `predicate=within` in which case it is flipped so that we can use contains (`tree_geom.contains(input_geom)`).

Let me know if there is anything else I missed!

jorisvandenbossche commented on geopandas/sindex.py:
```diff
@@ -81,22 +250,135 @@ def is_empty(self):
+    from . import geoseries  # noqa
+    from .array import GeometryArray, _shapely_to_geom  # noqa
+    import pygeos  # noqa
+    from pygeos import STRtree, box, points, Geometry  # noqa
```
I don't think `Geometry` is actually used below? 

Also, maybe we should just use the `import pygeos` for all usage of pygeos, that might be more explicit

adriangb replied:
Will do, I agree.

adriangb commented on benchmarks/sindex.py:
```diff
@@ -49,3 +50,44 @@ def time_sindex_intersects(self, *args):
+    def setup(self, *args):
+        triangles = GeoSeries(
+            [
+                Polygon([(random.random(), random.random()) for _ in range(3)])
```
I had copied that from the sindex benchmarks, I did not give it much thought. How about using the `naturalearth_lowres` dataset as the tree and `naturalearth_cities` as the input geometries? I think that's a pretty realistic use case.

jorisvandenbossche replied:
> I had copied that from the sindex benchmarks, I did not give it much thought, so thank you for checking. How about using the naturalearth_lowres dataset as the tree and naturalearth_cities as the input geometries?

That's fine I think. We can worry about good benchmarks cases (that maybe also cover different types of uses cases) later

adriangb commented on geopandas/sindex.py:
```diff
@@ -336,13 +342,15 @@ def query(self, geometry, predicate=None, sort=False):
-                If predicate is provided, the input geometry is tested using the predicate function against each
-                item in the tree whose extent intersects the envelope of the input geometry:
-                predicate(input_geometry, tree_geometry).
+            predicate : {None, 'intersects', 'within', 'contains', \
+                'overlaps', 'crosses', 'touches'}, optional
```
Fixed.

adriangb commented on geopandas/sindex.py:
```diff
@@ -286,9 +290,11 @@ def query_bulk(self, geometry, predicate=None, sort=False):
-            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
-                If predicate is provided, the input geometry is tested using the predicate function against each
-                item in the index whose extent intersects the envelope of the input geometry:
+            predicate : {None, 'intersects', 'within', 'contains', \
+                'overlaps', 'crosses', 'touches'}, optional
```
Fixed.

adriangb commented on geopandas/sindex.py:
```diff
@@ -209,11 +211,13 @@ def query_bulk(self, geometry, predicate=None, sort=False):
-                item in the tree whose extent intersects the envelope of the each input geometry:
-                predicate(input_geometry, tree_geometry).  If possible, prepared geometries are used to help
-                speed up the predicate operation.
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', \
+                'crosses', 'touches'}, optional
```
Fixed.

adriangb commented on geopandas/sindex.py:
```diff
@@ -108,20 +108,23 @@ def query(self, geometry, predicate=None, sort=False):
-                If predicate is provided, the input geometry is tested using the predicate function against each
-                item in the tree whose extent intersects the envelope of the input geometry:
-                predicate(input_geometry, tree_geometry).  If possible, prepared geometries are used to help
+            predicate : {None, 'intersects', 'within', 'contains', \
+                'overlaps', 'crosses', 'touches'}, optional
```
Interesting, this is for readthedocs to pick it up correctly?

brendan-ward commented on geopandas/tests/test_sindex.py:
```diff
@@ -172,10 +174,284 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+            (None, (0, 0, 1, 1), [[0, 0], [0, 1]]),
+            ("intersects", (-1, -1, -0.5, -0.5), [[], []]),
+            ("intersects", (-0.5, -0.5, 0.5, 0.5), [[0], [0]]),
+            ("intersects", (0, 0, 1, 1), [[0, 0], [0, 1]]),
```
Also add test cases for multiple input geometries; ideally 2 variants:
- both geometries intersect
- only second geometry interesects

adriangb replied:
Added these two types of tests.

brendan-ward commented on geopandas/tests/test_sindex.py:
```diff
@@ -172,10 +174,284 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+        """
+        # pass through GeoSeries to have GeoPandas
+        # determine if it should use shapely or pygeos geometry objects
+        test_geom = geopandas.GeoSeries([box(*test_geom)], index=["0"])
```
The implementation doesn't seem to match comment.  Using `box` here always returns a shapely geometry, right?

Suggestion: loop over both shapely and pygeos box methods here, so that all input geometries above are tested with each shapely and pygeos geometry types.

adriangb replied:
The idea is to let the other parts of `GeoPandas` determine what internal representation should be used. By creating a `GeoSeries`, the geometries will be converted to pygeos geometries automatically if `compat.USE_PYGEOS==True`. This way we don't have to add another check for the flag here, which might change in the future.

jorisvandenbossche replied:
Yes, no need to separately test with pygeos geometries, tests can just use shapely geometries. We run the all tests twice on CI, once with and once without PYGEOS enabled (which will automatically ensure that shapely geometries get converted to pygeos internally, if that is enabled)

brendan-ward commented on geopandas/tests/test_sindex.py:
```diff
@@ -172,10 +174,284 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+        res = self.df.sindex.query_bulk(test_geoms)
+        assert_array_equal(res, expected_value)
+
+    @pytest.mark.parametrize(
```
No need for `parametrize` here, additional variants are unlikely.

adriangb replied:
I tend to like keeping the `parametrize` because it makes the test cases more explicit instead of burying them into the test function body. Is there is another way to do this for a single testcase?

brendan-ward replied:
It's usually just a matter of personal preference, at that level.  Personally I fully expect that the details of the test are contained in the test function, unless there is a need to iterate over several input parameters.  Having it `parametrize` is a hint that the implementation should be tested over many variants, not having `parametrize` is a hint that a single variant is sufficient.

adriangb replied:
Ok, @jorisvandenbossche seemed to be of the same opinion as you; I went ahead and removed the parametrizations for single inputs.

adriangb commented:
@brendan-ward those parts of the rtree implementation are being tested, CI is just missing `--cov-append `, I'll add that :)

adriangb commented on geopandas/sindex.py:
```diff
@@ -81,22 +250,135 @@ def is_empty(self):
             self.objects = self.ids = original_indexes[~geometry.values.is_empty]
             super().__init__(non_empty.values.data)
 
+        def query_bulk(self, geometry, predicate=None, sort=False):
```
Agreed, will do.

adriangb commented:
The CI coverage fix (adding `--cov-append`) worked, diff coverage is now at 100%!

I added some tests as described above. @brendan-ward let me know if you think these are reasonable.

I also removed the unused `parametrize`.

adriangb commented:
Here are benchmarks for this latest version:
| pygeos      | rtree      | ratio  | test                                                 |
|-------------|------------|--------|------------------------------------------------------|
| 324±5μs     | 9.33±2ms   | 28.84  | sindex.Bench.time_sindex_index_creation              |
| 11.0±0.6ms  | 20.0±5ms   | ~1.82  | sindex.Bench.time_sindex_intersects                  |
| 48.5±0.9ms  | 22.4±6ms   | 0.46   | sindex.Bench.time_sindex_intersects_objects          |
| 57.5±0.8ms  | 107±20ms   | 1.86   | sindex.BenchQuery.time_query('contains')             |
| 3.12±0.03ms | 108±20ms   | 34.42  | sindex.BenchQuery.time_query('crosses')              |
| 59.3±0.8ms  | 34.6±6ms   | 0.58   | sindex.BenchQuery.time_query('intersects')           |
| 57.9±0.3ms  | 117±20ms   | 2.02   | sindex.BenchQuery.time_query('overlaps')             |
| 58.5±1ms    | 38.2±9ms   | 0.65   | sindex.BenchQuery.time_query('touches')              |
| 743±9μs     | 31.5±5ms   | 42.43  | sindex.BenchQuery.time_query('within')               |
| 488±5μs     | 33.2±5ms   | 68.02  | sindex.BenchQuery.time_query(None)                   |
| 56.3±0.7ms  | 92.3±1ms   | 1.64   | sindex.BenchQuery.time_query_bulk('contains')        |
| 2.59±0.03ms | 94.1±2ms   | 36.32  | sindex.BenchQuery.time_query_bulk('crosses')         |
| 55.3±0.1ms  | 30.7±0.6ms | 0.56   | sindex.BenchQuery.time_query_bulk('intersects')      |
| 56.3±0.8ms  | 94.8±5ms   | 1.69   | sindex.BenchQuery.time_query_bulk('overlaps')        |
| 55.7±0.2ms  | 33.1±0.5ms | 0.59   | sindex.BenchQuery.time_query_bulk('touches')         |
| 260±20μs    | 25.8±1ms   | 99.44  | sindex.BenchQuery.time_query_bulk('within')          |
| 235±3μs     | 29.0±0.2ms | 123.34 | sindex.BenchQuery.time_query_bulk(None)              |
| 55.4±0.1ms  | 104±40ms   | 1.87   | sindex.BenchQuery.time_query_bulk_data('contains')   |
| 2.56±0.02ms | 102±6ms    | 39.88  | sindex.BenchQuery.time_query_bulk_data('crosses')    |
| 56.8±2ms    | 32.1±0.4ms | 0.57   | sindex.BenchQuery.time_query_bulk_data('intersects') |
| 57.5±0.8ms  | 91.5±0.6ms | 1.59   | sindex.BenchQuery.time_query_bulk_data('overlaps')   |
| 55.8±0.8ms  | 32.9±0.4ms | 0.59   | sindex.BenchQuery.time_query_bulk_data('touches')    |
| 247±20μs    | 24.1±0.6ms | 97.53  | sindex.BenchQuery.time_query_bulk_data('within')     |
| 222±5μs     | 28.7±0.9ms | 129.50 | sindex.BenchQuery.time_query_bulk_data(None)         |


It's interesting that rtree is looking slightly faster for some tests, contrary to the previous benchmarks. Maybe the individual operations are faster, so if the number of matches is small the non-vectorized operations win?

adriangb commented:
I found a possible bug: for both implementations, there are crashes/errors when non-valid geometries are used. I see three options here:

1. Don't even check. Likely the fastest, but could result in confusion (with pygeos, the error message is internal).
2. Check and raise a useful error.
3. Check and skip invalid geometries (we already do this for empty geometries).

Let me know what you think. All should be relatively easy to implement.

jorisvandenbossche commented:
How do you get an error for the rtree implementation with invalid geometries? (does getting the bounds error?)

We shouldn't skip invalid geometries, IMO those are the responsibility of the user. It would be good that there is a decent error message of course. But ideally that can happen without having to check for them (at least on this level of the sindex).

> with pygeos, the error message is internal).

The that sounds something to solve in pygeos. Can you open an issue there with an example?

adriangb commented:
> How do you get an error for the rtree implementation with invalid geometries? (does getting the bounds error?)
> 
> We shouldn't skip invalid geometries, IMO those are the responsibility of the user. It would be good that there is a decent error message of course. But ideally that can happen without having to check for them (at least on this level of the sindex).
> 
> > with pygeos, the error message is internal).
> 
> The that sounds something to solve in pygeos. Can you open an issue there with an example?

I got the error during the predicate check I think. Probably from trying to check if something is contained in a self touching polygon or something like that. Let me try to make a simpler geometry than the ones in the naturalearth dataset and test.

jorisvandenbossche commented:
Ah, yes, I forgot the predicate check. In principle, it's "just" the predicate function that should raise an informative error message. And then we don't need to care about those here in the spatial index.

adriangb commented:
> Ah, yes, I forgot the predicate check. In principle, it's "just" the predicate function that should raise an informative error message. And then we don't need to care about those here in the spatial index.

Right, I think we should make pygeos give a more informative error to let the user know that there was an issue checking a predicate, and give the geometry index or something. I can open an issue there for that.

adriangb commented:
So back to this PR. Here are the things I think may need double checking:
- [ ]  Make sure the tests are covering all corner cases.
- [ ]  Make sure the benchmarks make sense and we don't have any unnecessary slowdowns.

jorisvandenbossche commented:
> I'm not even going to open an issue in pygeos because this is not a real world use case. Sorry for the distraction!

That's still welcome I would say. Invalid geometries are already an annoying part of daily life for gis analysts / geopandas users, so we should make sure it is not even more confusing with bad error messages.

adriangb commented:
> That's still welcome I would say. Invalid geometries are already an annoying part of daily life for gis analysts / geopandas users, so we should make sure it is not even more confusing with bad error messages.

Ok, posted the issue pygeos/pygeos#139

jorisvandenbossche commented:
Back to the actual PR. Regarding the tests, that would be nice to check (but I have the impression there is already quite a good coverage?). For the benchmarks, I think that is fine to leave as a follow-up.

adriangb commented:
> Regarding the tests, that would be nice to check (but I have the impression there is already quite a good coverage?). 

I personally think we are covering what we need to, but there is certainly stuff that we are supporting but not testing. For example, we are not testing a lot of predicates (ex: "touches"). I also don't want to just duplicate all of the testing within pygeos if we are just wrapping it. But I leave it up to you and others to decide how much testing we want.

> For the benchmarks, I think that is fine to leave as a follow-up.

Agreed, they are pretty informative for now, but it's probably worth discussing performance in detail once we start using `query_bulk` in other parts.

brendan-ward commented on geopandas/tests/test_sindex.py:
```diff
@@ -172,10 +181,299 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+            (None, box(-1, -1, -0.5, -0.5), []),
+            (None, box(-0.5, -0.5, 0.5, 0.5), [0]),
+            (None, box(0, 0, 1, 1), [0, 1]),
+            # bbox intersects but not geometry
```
This comment doesn't seem to explain the following cases.  

It might make these easier to group mentally by the input geometries instead of by predicate, so that it is easier to see the variations in output based on just the predicate:

```
(None, LineString([(0, 1), (1, 0)]), [0, 1])  # bounding box intersects
("intersects", LineString([(0, 1), (1, 0)]), []),  # geometry does not intersect
("within", LineString([(0, 1), (1, 0)]), []),  # intersects but not within
("contains", LineString([(0, 1), (1, 0)]), []),  # intersects but not contains
```

Might be good to add a `touches` variant to this one too, but overall I agree with other comments that predicate testing does not need to be exhaustive here if appropriately covered already for rtree implementation (since most predicates are at least partly covered by pygeos tests).

adriangb replied:
My comments were badly organized, that was supposed to refer to just the geometry under it. I added more comments and reorganized them.

I think I am going to skip reorganizing by input geometry. I do see the argument for it, but it would require reorganizing all test cases to be consistent. Let me know if this is okay.

brendan-ward replied:
It is fine not to reorganize, so long as the comments make sense for the conditions.

adriangb replied:
Ok. Let me know if the new comments help.

brendan-ward commented on geopandas/tests/test_sindex.py:
```diff
@@ -172,10 +181,299 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+                "contains",
+                LineString([(0, 1), (1, 0)]),
+                [],
+            ),  # intersects but not contained
```
Typo  `contained` => `contains`?

adriangb replied:
Fixed

brendan-ward commented on geopandas/tests/test_sindex.py:
```diff
@@ -172,10 +181,299 @@ def test_intersection_invalid_bounds_tuple(self, test_geom):
+        "predicate, test_geom, expected",
+        (
+            (None, (-1, -1, -0.5, -0.5), [[], []]),
+            ("intersects", (-1, -1, -0.5, -0.5), [[], []]),
```
Since multiple paths through the implementation can produce empty result arrays, it probably would be a good idea to include a case with a non-empty result too.

adriangb replied:
Added a case `("contains", (-1, -1, 1, 1), [[0], [0]])`

adriangb commented:
Thanks @brendan-ward. I will address the specific comments soon. A general question, regarding testing of predicates: we are currently using the same tests for both rtree and pygeos, this means there are predicates we are not testing with rtree.

The predicates we are testing are `intersects`, `within`, `contains` and `None`. We are not testing `overlaps`, `crosses` or `touches`.

All of the missing predicates share the same logic, so do you think we can get away with testing a single one (ex: `touches`)? I think this is what you are implying in your comments, but I just want to clarify.

adriangb commented:
> It might be good to add a test case that uses the countries / capitals, with varying predicates, to verify that the size of the results is the same between rtree and pygeos.

Are you suggesting we directly compare (dynamically setting the flag within the testcase or something), or that we hardcode the expected output size into the tests?

brendan-ward commented:
> Are you suggesting we directly compare (dynamically setting the flag within the testcase or something), or that we hardcode the expected output size into the tests?

Hardcode the expected size.

Using `touches` is fine.  The goal here is less about the correctness of the predicates, and more that we're correctly catching where they differ from the bounding box intersection.  More tests can always get added later, you've done a lot here!

adriangb commented:
I'll add an integration test with hardcoded sizes for the `naturalearth` dataset. And then I think we should be done?

adriangb commented:
Tests added. Let's get this merged!

brendan-ward commented on geopandas/tests/test_sindex.py:
```diff
@@ -504,3 +504,30 @@ def test_is_empty(self):
+        """Tests output sizes for the naturalearth datasets."""
+        world = read_file(datasets.get_path("naturalearth_lowres"))
+        capitals = read_file(datasets.get_path("naturalearth_cities"))
+        # Reproject to Mercator (after dropping Antartica)
```
Is reprojection necessary?  I would assume that geographic coordinates should work fine for this test.

adriangb replied:
I removed the reprojection and tested, all working. It is nicer to have less things being tested, so I think this is a good idea.

brendan-ward commented on geopandas/tests/test_sindex.py:
```diff
@@ -504,3 +504,30 @@ def test_is_empty(self):
+        capitals = capitals.to_crs("epsg:3395")[["geometry"]]
+
+        res = countries.sindex.query_bulk(capitals.geometry, predicate)
+        print(res.shape)
```
Remove unnecessary print statement.

adriangb replied:
Ups, good catch.

adriangb replied:
Fixed

adriangb commented:
Pinging @jorisvandenbossche and @martinfleis to re-review or merge if ready.

jorisvandenbossche commented:
@adriangb sorry for the delay, and thanks a lot for this PR !!

adriangb commented:
No problem, excited to move on to actually implementation this stuff now!
